### ISSUE TITLE
Rounding issue using BigDecimal/DateTime with Rails 6

### ISSUE THREAD

**Issue by ledermann** [rails/rails#38041]
I stumbled upon a rounding issue with Rails 6 using DateTime and BigDecimal. Everything is fine with Rails 5. Please try these examples in the Rails console:

```ruby
# Rails 5.2.4.1: OK
"2019-12-19 10:00".to_datetime - BigDecimal(1).hour
=> Thu, 19 Dec 2019 09:00:00 +0000
# OK

# Rails 6.0.2.1: WRONG
"2019-12-19 10:00".to_datetime - BigDecimal(1).hour
=> Thu, 19 Dec 2019 08:59:59 +0000
```

In Rails 6 there is a 1sec difference, which is not correct.

Both using Ruby 2.6.5

**Comment by daniel-rikowski** [comment]
Can confirm. This syntax variant doesn't work, too:

```ruby
"2019-12-19 10:00".to_datetime.advance(hours: -1.to_d)
=> Thu, 19 Dec 2019 08:59:59 +0000
```

But these are ok:

```ruby
"2019-12-19 10:00".to_datetime.advance(hours: 1.to_d)
=> Thu, 19 Dec 2019 11:00:00 +0000
"2019-12-19 10:00".to_datetime.advance(hours: -2.to_d)
=> Thu, 19 Dec 2019 08:00:00 +0000
"2019-12-19 10:00".to_datetime.advance(hours: -1.to_f)
=> Thu, 19 Dec 2019 09:00:00 +0000
```

**Comment by st0012** [comment]
This issue was introduced by https://github.com/rails/rails/commit/c85e3f65f3409fc329732912908c3601d8e5fac9

I've attached a fix for it.

**Comment by st0012** [comment]
This is actually a Ruby's [bug](https://bugs.ruby-lang.org/issues/16518). 

If the issue keeps bothering you, here's a quick fix


```ruby
class DateTime
  def since(seconds)
    self + Rational(Rational(seconds), 86400)
  end
end
```

**Comment by ledermann** [comment]
It seems that this bug was fixed in Ruby 3.0:
https://github.com/ruby/ruby/pull/3702

Using Ruby 3.0.2 (and Rails 6.1.4), it works fine now:

```ruby
$ ruby -v
ruby 3.0.2p107 (2021-07-07 revision 0db68f0233) [x86_64-darwin20]

$ rails c
Loading development environment (Rails 6.1.4)
> "2019-12-19 10:00".to_datetime - BigDecimal(1).hour
=> Thu, 19 Dec 2019 09:00:00 +0000 # OK
```
